[utils] Implemented Enumerate iterator adapter (#28)

derzhavin3016 opened the pull request:
Closes #28

derzhavin3016 committed:
[utils] Implemented Enumerate iterator adapter (#28)

Tako-San commented on include/utils/utils.hh:
```diff
@@ -0,0 +1,175 @@
+  {
+    std::size(cont)
+    } -> std::convertible_to<std::size_t>;
+  {
+    typename std::remove_reference_t<T>::iterator()
+    } -> std::forward_iterator;
```
Won't work if iterator class named as `Iterator` for example

derzhavin3016 replied:
Fixed

Tako-San commented on include/utils/utils.hh:
```diff
@@ -0,0 +1,175 @@
+template <ItContainer C>
+class Enumerate final
+{
+private:
+  C &cont_;
+  using EnumItType = detail::EnumerIt<typename std::remove_reference_t<C>::iterator>;
+
+public:
+  Enumerate(C &cont) : cont_(cont)
+  {}
```
I think it is better to define deleted operator=

derzhavin3016 replied:
I don't think it is necessary, because it has already deleted implicitly because of reference member

derzhavin3016 replied:
In case of storing by value, default `operator=` does the right thing
Of course there will be a deep copy for copy `operator=`, but I don't think it's also a reason to delete it

derzhavin3016 committed:
[utils] Improved move semantics of enumerate

- Removed class template specialization for rvalue reference by specifying storage type of container conditionally
- Added tests for checking move semantic's work
- Removed requirement of ::iterator type in container

derzhavin3016 committed:
[utils] Added const to enumerate iterator's count

- To prevent changing of counter, made it const in reference & value type
- Removed useless dummy_ variable because of errors while using it

derzhavin3016 committed:
[utils] Removed makeEnumerate

- Replaced makeEnumerate w/ ctor with deduction hint
- Added size method for enumerate

derzhavin3016 committed:
[utils] Added test + some fixes

- Added test with constant container
- Removed const from reference & value_type
- Removed operator!= for EnumerateIt (generated by default)
- Renamed Enumer -> Enumerate

derzhavin3016 commented:
I think we can merge it after I add test for ordinary array as a container

Tako-San commented on include/utils/utils.hh:
```diff
@@ -0,0 +1,137 @@
+
+public:
+  using iterator_category = std::forward_iterator_tag;
+  using difference_type =
+    std::pair<std::make_signed_t<std::size_t>, typename std::iterator_traits<It>::difference_type>;
```
Is it really should be a pair? Maybe just number is enough?

Tako-San commented on test/unit/utils.cc:
```diff
@@ -0,0 +1,188 @@
+  auto it1 = enumerate.begin();
+  auto it2 = enumerate.begin();
+
+  using value_type =
+    std::iterator_traits<detail::EnumerateIt<std::vector<int>::iterator>>::value_type;
+
+  value_type v1 = *it1++;
+  value_type v2 = *it1;
```
Why not auto?
